### PR TITLE
Update doc - tests - and conf template to reflect logic around recover_after_data_nodes

### PR DESCRIPTION
**Pull request by hlcianfagna**

### CODE REVIEW

File: docs/admin/system-information.rst [hunk 1]
```diff
@@ -834,9 +834,16 @@ Recovery expected data nodes
...
 :ref:`gateway.expected_data_nodes <gateway.expected_data_nodes>` setting checks
-that the number of data nodes that should be waited for the immediate
-cluster state :ref:`recovery <gloss-shard-recovery>`. This value must be equal
-to the maximum number of data nodes in the cluster.
+if the setting match the actual number of data nodes present in the cluster.
+The value of this settings is checked against 
+:ref:`gateway.recover_after_data_nodes
+<gateway.recover_after_data_nodes>` (or 
+:ref:`gateway.recover_after_nodes <gateway.recover_after_nodes>`
+) during startup to decide if the cluster can proceed with
```
Comment (hammerhead) [inline docs/admin/system-information.rst:842]: I think the closing bracket needs to be on the previous line, otherwise, this renders an additional space:
Screenshot 2022-12-30 at 14 31 57

File: docs/admin/system-information.rst [hunk 2]
```diff
@@ -834,9 +834,16 @@ Recovery expected data nodes
 
 The check for the
 :ref:`gateway.expected_data_nodes <gateway.expected_data_nodes>` setting checks
-that the number of data nodes that should be waited for the immediate
-cluster state :ref:`recovery <gloss-shard-recovery>`. This value must be equal
-to the maximum number of data nodes in the cluster.
+if the setting match the actual number of data nodes present in the cluster.
```
Comment (hammerhead) [inline docs/admin/system-information.rst:837]:
```suggestion
if the setting matches the actual number of data nodes present in the cluster.
```

File: docs/admin/system-information.rst [hunk 3]
```diff
@@ -834,9 +834,16 @@ Recovery expected data nodes
...
+The value of this settings is checked against 
+:ref:`gateway.recover_after_data_nodes
+<gateway.recover_after_data_nodes>` (or 
+:ref:`gateway.recover_after_nodes <gateway.recover_after_nodes>`
+) during startup to decide if the cluster can proceed with 
+:ref:`recovery <gloss-shard-recovery>`.
+If the actual number of nodes is below the number expected this is flagged to
```
Comment (hammerhead) [inline docs/admin/system-information.rst:844]:
```suggestion
If the actual number of nodes is below the expected number of nodes, this is flagged to
```

File: docs/admin/system-information.rst [hunk 4]
```diff
@@ -834,9 +834,16 @@ Recovery expected data nodes
...
+indicate some node[s] are down. If the actual number is greater this is
```
Comment (hammerhead) [inline docs/admin/system-information.rst:845]: I think you can drop the "[s]", otherwise you would also need to use "is/are" later in the sentence.
```suggestion
indicate some nodes are down. If the actual number is greater, this is 
```

File: docs/config/cluster.rst [hunk 5]
```diff
@@ -1259,15 +1259,13 @@ The following settings can be used to configure the behavior of the
   | *Default:*   ``-1``
   | *Runtime:*  ``no``
 
-  The setting ``gateway.expected_nodes`` defines the number of nodes that
-  should be waited for until the cluster state is recovered immediately. The
-  value of the setting should be equal to the number of nodes in the cluster,
-  because you only want the cluster state to be recovered after all nodes are
-  started.
+  The setting ``gateway.expected_nodes`` defines the total number of nodes 
+  expected in the cluster. It is evaluated together with recover_after_nodes
```
Comment (hammerhead) [inline docs/config/cluster.rst:1263]:
```suggestion
  expected in the cluster. It is evaluated together with ``recover_after_nodes``
```

File: docs/config/cluster.rst [hunk 6]
```diff
@@ -1259,15 +1259,13 @@ The following settings can be used to configure the behavior of the
...
+  to decide if the cluster can proceed with recovery.
```
Comment (hammerhead) [inline docs/config/cluster.rst:1264]:
```suggestion
  to decide if the cluster can start with recovery.
```
Not sure, is any recovery happening already before reaching `gateway.expected_nodes`?

File: docs/config/cluster.rst [hunk 7]
```diff
@@ -1276,11 +1274,10 @@ The following settings can be used to configure the behavior of the
...
   | *Runtime:*  ``no``
 
-  The setting ``gateway.expected_data_nodes`` defines the number of data nodes
-  that should be waited for until the cluster state is recovered immediately.
-  The value of the setting should be equal to the number of data nodes in the
-  cluster, because you only want the cluster state to be recovered after all
-  data nodes are started.
+  The setting ``gateway.expected_data_nodes`` defines the total number of
+  data nodes expected in the cluster. It is evaluated together with 
+  recover_after_data_nodes
```
Comment (hammerhead) [inline docs/config/cluster.rst:1279]:
```suggestion
  ``gateway.recover_after_data_nodes``
```

File: docs/concepts/clustering.rst [hunk 8]
```diff
@@ -151,8 +151,8 @@ across the cluster.
    versions 4.x and above `determine quorum size automatically
    <https://crate.io/docs/crate/howtos/en/latest/clustering/multi-node-setup.html#master-node-election>`_.
 
-A minimum number of nodes (referred as a *quorum*) needs to configured (using the
-`discovery.zen.minimum_master_nodes`_ setting) to ensure that in case of a
+A minimum number of nodes (referred as a *quorum*) needs to be configured
+to ensure that in case of a
```
Comment (seut) [inline docs/concepts/clustering.rst:155]: Why do we want to remove the mentioning of the `minimum_master_nodes` here? As stated above, this paragraph is only needed for CrateDB < 4 where it is important to know and understand the `minimum_master_nodes` setting.

File: server/src/test/java/io/crate/expression/reference/sys/check/node/SysNodeChecksTest.java [hunk 9]
```diff
@@ -230,7 +230,7 @@ public void testRecoveryAfterNodesCheckWithCorrectSetting() {
 
         var settings = Settings.builder()
             .put(GatewayService.RECOVER_AFTER_DATA_NODES_SETTING.getKey(), 2)
-            .put(GatewayService.EXPECTED_DATA_NODES_SETTING.getKey(), 2)
+            .put(GatewayService.EXPECTED_DATA_NODES_SETTING.getKey(), 3)
```
Comment (seut) [inline server/src/test/java/io/crate/expression/reference/sys/check/node/SysNodeChecksTest.java:233]: How is this change relevant to the documentation change? If not I prefer to remove it.

File: docs/concepts/clustering.rst (hunk 8 above)
Comment (seut) [reply docs/concepts/clustering.rst:155]: Or maybe we remove this section completely or replace it with the content described at https://crate.io/docs/crate/howtos/en/latest/clustering/multi-node-setup.html#master-node-election?

File: server/src/test/java/io/crate/expression/reference/sys/check/node/SysNodeChecksTest.java (hunk 9 above)
Comment (hlcianfagna) [reply server/src/test/java/io/crate/expression/reference/sys/check/node/SysNodeChecksTest.java:233]: My thinking here was that I was trying to clarify that it was legit to bring the cluster online with a minority of data nodes offline, so with the change we would now be testing if cluster checks fail on a scenario with 3 data nodes expected but recover_after_data_nodes set to 2

Comment (seut) [reply server/src/test/java/io/crate/expression/reference/sys/check/node/SysNodeChecksTest.java:233]: Ah ok, fine. Thanks for the clarification.

File: docs/concepts/clustering.rst (hunk 8 above)
Comment (hlcianfagna) [reply docs/concepts/clustering.rst]: I replaced the section with a note and a link, please take a look to see if it looks ok like this.

**Review by seut — APPROVED** [review]
Lgtm. Please don't forget to squash the commits before merging, thx!

File: docs/admin/system-information.rst [hunk 10]
```diff
@@ -834,9 +834,16 @@ Recovery expected data nodes
 
 The check for the
 :ref:`gateway.expected_data_nodes <gateway.expected_data_nodes>` setting checks
-that the number of data nodes that should be waited for the immediate
-cluster state :ref:`recovery <gloss-shard-recovery>`. This value must be equal
-to the maximum number of data nodes in the cluster.
+if the setting matches the actual number of data nodes present in the cluster.
+The value of this settings is checked against
```
Comment (hammerhead) [inline docs/admin/system-information.rst:838]:
```suggestion
The value of this setting is checked against 
```

**Review by hammerhead — APPROVED** [review]

**Review by proddata — CHANGES_REQUESTED** [review]
As said previously: The part about the strong requirement ("must") for at least half the expected data nodes is imho misleading.

File: docs/config/cluster.rst [hunk 11]
```diff
@@ -1276,16 +1275,15 @@ The following settings can be used to configure the behavior of the
...
+  The setting ``gateway.expected_data_nodes`` defines the total number of
+  data nodes expected in the cluster. It is evaluated together with 
+  ``gateway.recover_after_data_nodes``
+  to decide if the cluster can proceed with recovery.
 
 .. _gateway.recover_after_time:
 
 **gateway.recover_after_time**
-  | *Default:*   ``0ms``
+  | *Default:*   ``5m`` if ``expected_nodes`` or ``expected_data_nodes`` are set,  ``0ms`` otherwise
```
Comment (proddata) [inline docs/config/cluster.rst:1286]: This defaults to `5m`. There can't be 2 default values
 it is just not relevant  if no  expected nodes are set.

File: docs/config/cluster.rst [hunk 12]
```diff
@@ -1300,11 +1298,12 @@ The following settings can be used to configure the behavior of the
   | *Runtime:*  ``no``
 
   The ``gateway.recover_after_nodes`` setting defines the number of nodes that
-  need to be started before the cluster state recovery will start. Ideally the
-  value of the setting should be equal to the number of nodes in the cluster,
-  because you only want the cluster state to be recovered once all nodes are
-  started. However, the value must be bigger than the half of the expected
+  need to be started before the cluster state recovery will start.
+  The value must be bigger than half the expected
```
Comment (proddata) [inline docs/config/cluster.rst:1302]: Is this really a "must" or a "should"?

File: docs/sql/statements/create-table.rst [hunk 13]
```diff
@@ -683,17 +683,6 @@ Sets size of transaction log prior to flushing.
   Size (bytes) of translog.
 
 
-.. _sql-create-table-translog-interval:
```
Comment (proddata) [inline docs/sql/statements/create-table.rst:686]: Why do we remove this? This setting is relevant, if the translog durability is set to async

**Review by proddata — CHANGES_REQUESTED** [review]

File: docs/sql/statements/create-table.rst (hunk 13 above)
Comment (proddata) [reply docs/sql/statements/create-table.rst:686]: sorry, mistook this for the translog sync invertal

File: docs/config/cluster.rst (hunk 11 above)
Comment (seut) [reply docs/config/cluster.rst:1286]: No, it is like hernan described, the default value has a dependency.

File: docs/config/cluster.rst [hunk 14]
```diff
@@ -1319,10 +1318,11 @@ The following settings can be used to configure the behavior of the
...
   The ``gateway.recover_after_data_nodes`` setting defines the number of data
   nodes that need to be started before the cluster state recovery will start.
-  Ideally the value of the setting should be equal to the number of data nodes
-  in the cluster, because you only want the cluster state to be recovered once
-  all data nodes are started. However, the value must be bigger than the half
-  of the expected number of data nodes in the cluster.
+  The value must be bigger than half the expected
+  number of data nodes in the cluster.
+  Only set this to a number below the total number of data nodes
+  after checking the number of nodes required for quorum and the number of
```
[the remark this answers is not in the document]
Comment (seut) [reply docs/config/cluster.rst:1324]: I don't understand your example, how can the quorum be reached with 2 nodes started out of 100?

[the remark this answers is not in the document]
Comment (proddata) [reply docs/config/cluster.rst:1324]: You can have a quorum with 1 node in a 100 node cluster, with 1 node set to `node.master=true` and all others to `node.master=false`.

Only master nodes are relevant for a quorum. -> a setting that only checks for data nodes is completely irrelevant for any quorum.

[the remark this answers is not in the document]
Comment (seut) [reply docs/config/cluster.rst:1324]: I agree that this setting is not related to any quorum, but also not to any master node settings (that's why I was confused by your example).
I suggest that we keep the original description and just remove the "value must be bigger than the half of the expected number of data nodes" constrain.
```
The ``gateway.recover_after_data_nodes`` setting defines the number of data nodes
that need to have joined the cluster before the local shard recovery will start.
Ideally the value of the setting should be equal to the number of data nodes
in the cluster, because you only want the cluster state to be recovered once
all data nodes are started to avoid that shards are re-allocated
(replicas promoted to primary, new replicas are allocated).
```

[the remark this answers is not in the document]
Comment (proddata) [reply docs/config/cluster.rst:1324]: > Ideally the value of the setting should be equal to the number of data nodes

I slightly disagree with this statement. This is the role of `expected_data_nodes`. `recover_after_data_nodes` is a lower bound to still recover, if the later is not met. Maybe we can avoid recommendation - as they alway depend on the concrete setup - and rather focus on explaining what the setting does. E.g. in a multi-zone setup the argument for replicas is irrelevant, if zone awareness is enforced.

> you only want the cluster state to be recovered once all data nodes are started

Would also be careful with this, as data recovery is independent of cluster state recovery. We had big confusions around this topics (i.e. that cluster availability is somewhat independent from data availability)

[the remark this answers is not in the document]
Comment (seut) [reply docs/config/cluster.rst:1324]: @proddata Ah yes right, I was in the wrong setting context. Like you says, `recover_after_data_node` is the minimum threshold to wait with a timeout for the desired final state of `expected_data_nodes`. So +1 to remove any recommendation.

> Would also be careful with this, as data recovery is independent of cluster state recovery.

Also totally agreeing on this, it isn't about the cluster state recovery but shard recovery (data).
Let's remove any recommendation.

File: docs/config/cluster.rst (hunk 11 above)
Comment (proddata) [reply docs/config/cluster.rst:1286]: I think it is more complex then that ... Also the "default" would be `null` and not `0ms` , though being a technicality, it behaves slightly differently, then setting `0ms` explicitly.

```java
public static final TimeValue DEFAULT_RECOVER_AFTER_TIME_IF_EXPECTED_NODES_IS_SET = TimeValue.timeValueMinutes(5);
```

```java
if (RECOVER_AFTER_TIME_SETTING.exists(settings)) {
// explicitly set recover_after_time
    recoverAfterTime = RECOVER_AFTER_TIME_SETTING.get(settings);
} else if (expectedNodes >= 0 || expectedDataNodes >= 0 || expectedMasterNodes >= 0) {
// explicitly set expected nodes
    recoverAfterTime = DEFAULT_RECOVER_AFTER_TIME_IF_EXPECTED_NODES_IS_SET;
} else {
// default
    recoverAfterTime = null;
}
```

check later for `null`
```java
if (enforceRecoverAfterTime && recoverAfterTime != null) {
```

The setting is only relevant if
- set explicitly
_and/or_
- one of the expected nodes settings >= 1

How it behaves, is a combination of all 3 settings.
- `recover_after_nodes` are **always** a lower limit
- `expected_nodes` are an upper limit (unless `-1`) 
then: 
- if `expected_nodes` are set to `-1` (implicitly / explicitly)
  - if `recover_after_time` is not set implicitly => `null` -> immediate recover
  - if `recover_after_time` is set explicitly => recover after set time
- if `expected_nodes` are set to `0` or `1` 
   ->  immediate recovery independent (!) of setting
- if `expected_nodes` are set to  `>=1` 
  - if `recover_after_time` is not set implicitly => `5m` -> recover after `5m` if any node joins
  - if `recover_after_time` is set explicitly => recover after set time

---

```yaml
# recover_after_data_nodes: -1
# expected_data_nodes: -1
# recover_after_time: 5m
# --> recoverAfterTime == null
# --> recovers immediatily after any number of data nodes joined

recover_after_data_nodes: 2
# expected_data_nodes: -1
# recover_after_time: 5m
# --> recoverAfterTime == null
# --> recovers immediatily after 2 data nodes joined

# recover_after_data_nodes: -1
expected_data_nodes: -1
# recover_after_time: 5m
# --> recoverAfterTime == null
# --> recovers immediatily after any number of data nodes joined

# recover_after_data_nodes: -1
expected_data_nodes: 0
# recover_after_time: 5m
# --> recoverAfterTime == 5m
# --> recovers immediatily after any number of data nodes joined

# recover_after_data_nodes: -1
expected_data_nodes: 3
# recover_after_time: 5m
# --> recoverAfterTime == 5m
# --> recovers 5 minutes after any number of nodes joined
# --> recovers immediatily after 3 data nodes joined

# recover_after_data_nodes: -1
# expected_data_nodes: -1
recover_after_time: 3m
# --> recoverAfterTime == 3m
# --> recovers 3 minutes after any number of nodes joined

recover_after_data_nodes: 2
expected_data_nodes: 3
# recover_after_time: 5m
# --> recoverAfterTime == 5m
# --> recovers 5 minutes after 2 data nodes joined
# --> recovers immediatily after 3 data nodes joined

recover_after_data_nodes: 2
# expected_data_nodes: -1
recover_after_time: 3m
# --> recoverAfterTime == 3m
# --> recovers 3 minutes after 2 data nodes joined

# recover_after_data_nodes: 2
expected_data_nodes: 3
recover_after_time: 3m
# --> recoverAfterTime == 3m
# --> recovers 3 minutes after any number of nodes joined
# --> recovers immediatily after 3 data nodes joined

recover_after_data_nodes: 2
expected_data_nodes: 3
recover_after_time: 3m
# --> recoverAfterTime == 3m
# --> recovers 3 minutes after 2 data nodes joined
# --> recovers immediatily after 3 data nodes joined
```

**Comment by hlcianfagna** [comment]
I made some mistakes trying to rebase the branch on my GitHub Desktop.
I will continue this on PR13567 
I am taking on board all your comments, thank you for all the input, and apologies for the confusion.
